Documentation/botr/readytorun-overview.md

```diff
@@ -0,0 +1,335 @@
+
+The runtime currently supports two mechanisms for virtual dispatch. One mechanism is called virtual stub dispatch (VSD). It is used when calling interface methods. The other is a variation on traditional vtable-based dispatch and it is used when a non-interface virtual is called. We first discuss the VSD approach. 
+
+Assume that holds RCX holds a `Intf` then the call to `intfMethod()` would look like
+
+	     CALL [PTR_CALLSITE]
+	.data // in the data section
+    PTR_CALLSITE: INT_PTR = RUNTIME_ENTRY_FIXUP_METHOD // One per call SITE. 
+
+This looks same as the cross-module, non-virtual case, but there are important differences. Like the non-virtual case there is an indirect call through a pointer that lives in the module. However unlike the non-virtual case, there is one such slot per call site (not per target). What is in this slot is always guaranteed to get to the target (in this case to `Intf.intfMethod()`), but it is expected to change over time. It starts out pointing to a 'dumb' stub which simply calls a runtime helper that does the lookup (in likely a slow way. However, it can update the `PTR_CALLSITE` slot to a stub that efficiently dispatches to the interface for the type that actually occurred (the remaining details of stubbed based interface dispatch are not relevant to versioning). 
+
+The above description is accurate for the current CLR implementation for interface dispatch. What's more, is that nothing needs to be changed about the code generation to make it version resilient. It 'just works' today. Thus interface dispatch is version resilient with no performance penalty. 
+
+What's more, we can actually see VSD is really just a modification of the basic 'indirect call through updateable slot' technique that was used for non-virtual method dispatch. The main difference is that because the target depends on values that are not known until runtime (the type of the 'this' pointer), the 'fixup' function can never remove itself completely but must always check this runtime value and react accordingly (which might include fixing up the slot again. To make as likely as possible that the value in the fixup slot stabilizes, we create a fixup slot per call site (rather than per target).
+
+### Vtable Dispatch
+
+The CLR current also supports doing virtual dispatch through function tables (vtables). Unfortunately, vtables have the same version resilience problem as fields. This problem can be fixed in a similar way, however unlike fields, the likelihood of having many cross bubble fixups is higher for methods than for instance fields. Further, unlike fields we already have a version resilient mechanism that works (VSD), so it would have be better than that to be worth investing in. Vtable dispatch is only better than VSD for in polymorphic call sites (where VSD needs to resort to a hash lookup). If we find we need to improve dispatch for this case we have a some possible mitigations to try
```

Comment:
- ~~so it would have be better~~ so it would have *to* be better
- ~~better than VSD for in polymorphic~~ better than VSD for polymorphic
- ~~we have a some possible mitigations~~ we have some possible mitigations